### PR TITLE
tools/rados: add a parameter "--offset" to rados put command

### PR DESCRIPTION
**Pull request by Liuchang0812**
ref to #8269 , I have rebased the code, add a test case for it.

build pass and test pass.

@majianpeng @yangdongsheng @tchaikov need review please :smile:

### CODE REVIEW

File: qa/workunits/rados/test_rados_tool.sh [hunk 1]
```diff
@@ -534,12 +535,58 @@ function test_append()
   rm -rf ./rados_append_4k ./rados_append_4k_out ./rados_append_10m ./rados_append_10m_out
 }
 
+function test_put()
+{
+  # rados put test:
+  # replicated pool
+  $CEPH_TOOL osd pool create rados_put 100 100 replicated
```
Comment (tchaikov) [inline qa/workunits/rados/test_rados_tool.sh]: why not just use `$POOL`, which is already created in this test? if we do the cleanup right, there is no need to create another pool just for this test.

File: qa/workunits/rados/test_rados_tool.sh [hunk 2]
```diff
@@ -534,12 +535,58 @@ function test_append()
...
+
+  # create object
+  dd if=/dev/urandom of=rados_object_1k bs=1KB count=1
+
+  # test put command
+  $RADOS_TOOL -p rados_put put rados_put_obj ./rados_object_1k
+  $RADOS_TOOL -p rados_put get rados_put_obj ./rados_object_1k_out
+  diff ./rados_object_1k ./rados_object_1k_out > /dev/null
+  if [ $? == 0 ]; then
```
Comment (tchaikov) [inline qa/workunits/rados/test_rados_tool.sh]: no need to check the return value explicitly. we have `set -e` at line 269.

File: qa/workunits/rados/test_rados_tool.sh [hunk 3]
```diff
@@ -534,12 +535,58 @@ function test_append()
...
+  # replicated pool
+  $CEPH_TOOL osd pool create rados_put 100 100 replicated
+
+  # create object
+  dd if=/dev/urandom of=rados_object_1k bs=1KB count=1
+
+  # test put command
+  $RADOS_TOOL -p rados_put put rados_put_obj ./rados_object_1k
+  $RADOS_TOOL -p rados_put get rados_put_obj ./rados_object_1k_out
+  diff ./rados_object_1k ./rados_object_1k_out > /dev/null
```
Comment (tchaikov) [inline qa/workunits/rados/test_rados_tool.sh]: might want to use `cmp` to compare binary files.

File: qa/workunits/rados/test_rados_tool.sh [hunk 4]
```diff
@@ -534,12 +535,58 @@ function test_append()
...
+  if [ $? == 0 ]; then
+        echo "Both file are same"
+  else
+        die "Put with offset=0 failed expecting same file content"
+  fi
+
+  # test put command with offset 100
+  $RADOS_TOOL -p rados_put put rados_put_obj_offset_100 ./rados_object_1k --offset 100
+  $RADOS_TOOL -p rados_put get rados_put_obj_offset_100 ./rados_object_offset_100_out
+  dd if=rados_object_1k seek=100 bs=1c of=rados_object_offset_100_local > /dev/null
```
Comment (tchaikov) [inline qa/workunits/rados/test_rados_tool.sh]: better off using `cmp` which allows us to specify the bytes to skip for FILE1 and FILE2 separately.

File: qa/workunits/rados/test_rados_tool.sh [hunk 5]
```diff
@@ -534,12 +535,58 @@ function test_append()
...
+  diff ./rados_object_offset_100_local ./rados_object_offset_100_out > /dev/null
+  if [ $? == 0 ]; then
+        echo "Both file are same"
+  else
+        die "Put with offset=100 failed expecting same file content"
+  fi
+
+  $CEPH_TOOL osd pool delete rados_put rados_put --yes-i-really-really-mean-it
+  rm -rf ./rados_object_1k ./rados_object_1k_out ./rados_object_offset_0_out ./rados_object_offset_100_out
```
Comment (tchaikov) [inline qa/workunits/rados/test_rados_tool.sh]: could also reuse `$OBJ` and `cleanup` to perform this test.

File: qa/workunits/rados/test_rados_tool.sh (hunk 1 above)
Comment (Liuchang0812) [inline qa/workunits/rados/test_rados_tool.sh]: ok, i saw that test_append creates a new pool, I will try to fix those comments. Very thanks!

**Comment by Liuchang0812** [comment]
@tchaikov I fix those comments. I would like to fix similar problems in test_append/omap in another PR.

**Comment by Liuchang0812** [comment]
fixed test error

**Pull request merged**